### PR TITLE
Added support to washer-dryers to override some of the settings defined when a course is selected. Partial implementation of feature request #596

### PR DESCRIPTION
**Pull request by lancasterJ**
This PR provides entities for overriding the temperature, spin and rinse options set-up when a washer course is selected.  It checks that override values are permitted for the selected course and provide an error message if an invalid value is selected.  

I have developed this because I regularly use the same course with a different water temperature to the default.  I have tested this code using HA script and lovelace using a machine in the UK.  It should also work for a machine in a different region, but I cannot test this.

### CODE REVIEW

**Comment by lancasterJ** [comment]
@ollo69 I see some checks need to be run.  If you send me a pointer to the documentation I will try to run them.  Regards John

File: custom_components/smartthinq_sensors/wideq/devices/washerDryer.py [hunk 1]
```diff
@@ -124,6 +126,13 @@ def __init__(
         self._course_keys: dict[CourseType, str | None] | None = None
         self._course_infos: dict[str, str] | None = None
         self._selected_course: str | None = None
+        self._course_overrides: dict | None = {}
+        self._course_overrides_lists: dict | None = {
+            'temp': ['TEMP_COLD', 'TEMP_20', 'TEMP_30', 'TEMP_40', 'TEMP_60', 'TEMP_95'],
+            'spin': ['NO_SPIN', 'SPIN_400', 'SPIN_800', 'SPIN_1000', 'SPIN_Max'],
+            'rinse':  ['RINSE_NORMAL', 'RINSE_PLUS']}
+            # Need to initialise this list so that the UI can setup the select drop down.
+            # A better solution would be to generate this information from the data returned ThinQ API.
```
Comment (ollo69) [inline custom_components/smartthinq_sensors/wideq/devices/washerDryer.py:135]: This values can be different depending on device model and should be retrieved from model info. This could work for your device but not for other. I need to be sure that only supported modes can be used.

File: custom_components/smartthinq_sensors/wideq/devices/washerDryer.py [hunk 2]
```diff
@@ -708,6 +751,71 @@ async def select_start_course(self, course_name: str) -> None:
...
+
+    @property
+    def select_spin_enabled(self) -> bool:
+        """Return if select spin is enabled."""
+        return self._select_enabled('spin')
+        
+    async def select_start_spin(self, spin_name: str) -> None:
+        """Select a secific spin for remote start."""
+        self._select_start_option('spin', 'Spin Speed', spin_name)
+
```
Comment (ollo69) [inline custom_components/smartthinq_sensors/wideq/devices/washerDryer.py:818]: This will make all the logic very complex, because there are different combination that can vary depending on selected course. I would prefer to just add some parameter to the service call and allow to change this value calling the HA service with prober parameters, that will be ignored if not valid.

File: custom_components/smartthinq_sensors/select.py [hunk 3]
```diff
@@ -42,13 +42,41 @@ class ThinQSelectEntityDescription(
...
+    ),
+        ThinQSelectEntityDescription(
+        key="spin_selection",
+        name="Set Spin Speed",
+        icon="mdi:tune-vertical-variant",
+        options_fn=lambda x: x.device.spins_list,
+        select_option_fn=lambda x, option: x.device.select_start_spin(option),
+        available_fn=lambda x: x.device.select_spin_enabled,
+        value_fn=lambda x: x.device.selected_spin,
+    ),
```
Comment (ollo69) [inline custom_components/smartthinq_sensors/select.py:79]: As in my previous comment, I prefer to not use a select but just provide additional parameter to be used in the `remote_start` service.

**Review by ollo69 — CHANGES_REQUESTED** [review]

File: custom_components/smartthinq_sensors/wideq/devices/washerDryer.py (hunk 2 above)
Comment (lancasterJ) [reply custom_components/smartthinq_sensors/wideq/devices/washerDryer.py:818]: @ollo69 , 

> This will make all the logic very complex, because there are different combination that can vary depending on selected course

You are correct the ability to override a temp, spin or rinse depends on the WM model and the course selected.  The PR handles this by extracting the necessary data for course info and storing it in `_course_overrides_lists`.  The code should not required updating for different or future VM models unless a new setting is introduced (e.g. amount washing liquid to use) **and** we want to provide the user with a GUI selector to set it.  Background: In the GUI and API the temp, spin and rinse select elements are not enabled until a course is selected that permits them to be overridden.  For example, if the course Cotton is selected temp, spin and rinse are enabled, while for the course Allergy Care only spin and rinse are enabled.  For each enabled override, a list of allowed values is extracted from from the course description.  When a user attempts to set a value it is checked against this list.  If the value is not in the list it is ignored and the user is informed of the permitted values by raising `ServiceValidationError`.

> I would prefer to just add some parameter to the service call and allow to change this value calling the HA service with prober parameters, that will be ignored if not valid

When I started this work you had not publish the change to `remote_start` that allowed a course name string to be passed as a parameter.  I have since studied your change and learnt how an integration can define a service that can be used by an automation or script.

I feel it is very easy to add another parameter to `remote_start` that accepts a JSON dictionary of overrides.  The method can check `_course_overrides_lists` to confirm if the setting (e.g. spin speed) can be overridden and that the value is correct.  An example call would be `remote_start('Eco 40-60', '{"temp": "TEMP_60", "spin": "SPIN_400"}')` 

If you are happy with the above approach I am happy to modify the PR.  I feel it will be a simple change. If you are thinking of something different please let me know.

I feel there is one problem we should agree on.  For course names users can discover the valid course names that can be passed to `remote_start` using the course selection entity on the integration GUI.  There may be another method that I have not found.  How will a user, for a given course, discover setting that can be overridden and their permitted values that can be passed to `remote_start`. The selectors for temp, spin and rinse in the integration GUI will provide this information to the user.  If you don't wish to have these, how do you think we can expose the information in `_course_overrides_lists` to the user?

In summary:
1. I like the idea of adding overrides to `remote_start`.  It gives a cleaner HA script.
2. I believe the code will handle all current and future course settings and WM models.
3. Without a discovery mechanism for the overrides that is simple for the user I don't feel we have usable solution. I am therefore reluctant to remove the temp, spin and rinse selectors if we don't provide an alternative. Nevertheless, this is your call.

Regards
John

File: custom_components/smartthinq_sensors/select.py (hunk 3 above)
Comment (lancasterJ) [reply custom_components/smartthinq_sensors/select.py:79]: Please see my comments above.

File: custom_components/smartthinq_sensors/wideq/devices/washerDryer.py (hunk 1 above)
Comment (lancasterJ) [reply custom_components/smartthinq_sensors/wideq/devices/washerDryer.py]: @ollo69 

> This values can be different depending on device model and should be retrieved from model info. 

Now retrieved from model info.

**Comment by lancasterJ** [comment]
@ollo69 I have updated this PR to address the issues you raised.  I have not removed the selectors yet. I am waiting for your feedback to my question about a discovery mechanism.

**Comment by lancasterJ** [comment]
Hi @ollo69 
What do you require to move this PR forward?
Regards
John

**Comment by lancasterJ** [comment]
@ollo69  I have updated this PR to address the issues you have raised. Also, I have added service validation checks and messages that will tell a service user the course options available and permitted values for their machine in the local language.

I have not removed the selectors yet. If the selectors are the reason you will not take this PR then I will remove them.

**Comment by ollo69** [comment]
Unfortunately this year I'm very busy with my job and my time to work on this integration is very limited.
Please rebase this PR, I hope  that I will found the time to review this shortly 👍

**Comment by lancasterJ** [comment]
@ollo69, as requested I have re-based this PR.

**Comment by w-marco** [comment]
Hi, I have merged this locally as I wanted to try it out, but I get the following error:

```
Error while setting up smartthinq_sensors platform for select
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 361, in _async_setup_platform
    await asyncio.shield(awaitable)
  File "/config/custom_components/smartthinq-sensors/select.py", line 146, in async_setup_entry
    _async_discover_device(lge_cfg_devices)
  File "/config/custom_components/smartthinq-sensors/select.py", line 137, in _async_discover_device
    LGESelect(lge_device, select_desc)
  File "/config/custom_components/smartthinq-sensors/select.py", line 170, in __init__
    self._attr_options = self.entity_description.options_fn(self._api)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/smartthinq-sensors/select.py", line 57, in <lambda>
    options_fn=lambda x: x.device.temps_list,
                         ^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/smartthinq-sensors/wideq/backports/functools.py", line 68, in __get__
    val = self.func(instance)
          ^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/smartthinq-sensors/wideq/devices/washerDryer.py", line 269, in temps_list
    return list(self._option_friendly_names['temp']['enum_values'].values())
                ~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^
KeyError: 'temp'
```

My washer supports setting the temperature and course and such through the thinQ App and I could provide some mitmproxy JSON output if necessary for my model. Sadly I am not invested enough inside the wideq code to fix this right now.

**Comment by lancasterJ** [comment]
> Hi, I have merged this locally as I wanted to try it out, but I get the following error:
> 
> ```
> Error while setting up smartthinq_sensors platform for select
> Traceback (most recent call last):
>   File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 361, in _async_setup_platform
>     await asyncio.shield(awaitable)
>   File "/config/custom_components/smartthinq-sensors/select.py", line 146, in async_setup_entry
>     _async_discover_device(lge_cfg_devices)
>   File "/config/custom_components/smartthinq-sensors/select.py", line 137, in _async_discover_device
>     LGESelect(lge_device, select_desc)
>   File "/config/custom_components/smartthinq-sensors/select.py", line 170, in __init__
>     self._attr_options = self.entity_description.options_fn(self._api)
>                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/config/custom_components/smartthinq-sensors/select.py", line 57, in <lambda>
>     options_fn=lambda x: x.device.temps_list,
>                          ^^^^^^^^^^^^^^^^^^^
>   File "/config/custom_components/smartthinq-sensors/wideq/backports/functools.py", line 68, in __get__
>     val = self.func(instance)
>           ^^^^^^^^^^^^^^^^^^^
>   File "/config/custom_components/smartthinq-sensors/wideq/devices/washerDryer.py", line 269, in temps_list
>     return list(self._option_friendly_names['temp']['enum_values'].values())
>                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^
> KeyError: 'temp'
> ```
> 
> My washer supports setting the temperature and course and such through the thinQ App and I could provide some mitmproxy JSON output if necessary for my model. Sadly I am not invested enough inside the wideq code to fix this right now.

Hi @ollo69,
I think this is a timing race or a localization issue.  I have add a few lines of debug.  Please test again with debug enabled and send me the debug log.
FYI, I have re-based the the PR to v0.40.3.
Thanks
John

**Comment by w-marco** [comment]
Hi @lancasterJ ,

I have downloaded the latest version of your feature branch and loaded it into HA. 

My issue with select persisted, and that stopped the dryer from populating info, so I had to comment out the `return list...` inside wideq/devices/washerDryer.py. 

I also added the debug log to rinse and spin, in addition to course and temp. With this, I was able to get the debug logs, though of course the selects now don't work because I never return actual information to it.

This is the full debug log from my washer:
```
2024-10-31 08:48:54.181 DEBUG (MainThread) [custom_components.smartthinq_sensors.wideq.devices.washerDryer] _option_friendly_names: {'temp': {'friendly_name': 'Temp.', 'enum_values': {'NO_TEMP': 'Not Selected', 'TEMP_COLD': 'Kalt', 'TEMP_20': '20℃', 'TEMP_30': '30℃', 'TEMP_40': '40℃', 'TEMP_50': '50℃', 'TEMP_60': '60℃', 'TEMP_95': '95℃'}}, 'rinse': {'friendly_name': 'Spülen', 'enum_values': {'NO_RINSE': 'Not Selected', 'RINSE_NORMAL': 'Normal', 'RINSE_PLUS': 'Spülen+', 'RINSE_PLUSPLUS': 'Spülen++', 'RINSE_NORMAL_HOLD': 'Schleuderstopp', 'RINSE_PLUS_HOLD': 'Spülen+Schleuderstopp'}}, 'spin': {'friendly_name': 'Schleudern', 'enum_values': {'NOT_SELECTED': 'Not Selected', 'NO_SPIN': 'No', 'SPIN_400': '400', 'SPIN_600': '600', 'SPIN_700': '700', 'SPIN_800': '800', 'SPIN_900': '900', 'SPIN_1000': '1000', 'SPIN_1100': '1100', 'SPIN_1200': '1200', 'SPIN_1400': '1400', 'SPIN_1600': '1600', 'SPIN_Max': 'max.'}}}
```

**Comment by lancasterJ** [comment]
Hi @ollo69 

I cannot find away to sync machine info loading with setting up the selectors, so I have removed the selectors.  This means that the only way to override a default course setting is to use the SmartThinQ LGE Sensors: Remote Start action in a script.  Invalid parameters will be logged.

I recommend using Actions tab on the Developer Tools page for testing.  Invalid values for the Programme and Option Overrides will be reported on this screen. This give the user a way to discovery valid course and override combinations.

If this version works I need to do some code cleaning before you merge it.

The same race condition must exist with the course_list function.  Luckily, users are not seeing it.

Regards
John

**Comment by m-wack** [comment]
Hi @lancasterJ, 

I'm not the dev but still tried out the new changes as I reported the issues with select and such in the first place.
I am happy to report that the integration now works for me and I was successful in overriding the course parameters using the remote_start service call.

I used the following: Eco 40-60 with JSON {"temp": "TEMP_60", "spin": "SPIN_1400", "rinse": "RINSE_NORMAL"} and I was then able to remote start the washer with my desired program and parameters. 
This is great because I can finally set all the stuff in the morning automatically as my washer would sometimes lose my presets from the night before when I enabled remote start. 

Thank you for implementing this! 

@ollo69 I think once code-cleanup is done this would be a great addition to the integration and with a bit of proper documentation it could benefit lots of users. I would've still preferred the selects as they're more user-friendly but I think this will suffice for now.

**Comment by w-marco** [comment]
Hi @lancasterJ,

I'm not the dev but still tried out the new changes as I reported the issues with select and such in the first place.
I am happy to report that the integration now works for me and I was successful in overriding the course parameters using the remote_start service call.

I used the following: Eco 40-60 with JSON {"temp": "TEMP_60", "spin": "SPIN_1400", "rinse": "RINSE_NORMAL"} and I was then able to remote start the washer with my desired program and parameters.
This is great because I can finally set all the stuff in the morning automatically as my washer would sometimes lose my presets from the night before when I enabled remote start.

Thank you for implementing this!

@ollo69 I think once code-cleanup is done this would be a great addition to the integration and with a bit of proper documentation it could benefit lots of users. I would've still preferred the selects as they're more user-friendly but I think this will suffice for now.

**Comment by w-marco** [comment]
Hi!

Is there an update on this MR? I would really love to have this integrated.

**Comment by MPW1412** [comment]
Can we please have this merged?

**Comment by rddelacosta** [comment]
> 2024-10-31 08:48:54.181 DEBUG (MainThread) [custom_components.smartthinq_sen

Hi. What exactly is the service call that you use?

I've been trying the following but it does not work:
```
action: smartthinq_sensors.remote_start
target:
  entity_id: sensor.lg_washer
data: JSON {"course": "Delicates", "temp": "TEMP_30", "spin": "SPIN_1400", "rinse": "RINSE_NORMAL"}
```

and

`action: smartthinq_sensors.remote_start
target:
  entity_id: sensor.lg_washer
data:
  course: Delicates
  temp: TEMP_30
  spin: SPIN_1400
  rinse: RINSE_NORMAL`

It says "Failed to perform the action smartthinq_sensors.remote_start. extra keys not allowed @ data['rinse']. Got None extra keys not allowed @ data['spin']. Got None extra keys not allowed @ data['temp']. Got None"

I have followed the manual installation but referred to lancasterJ's git for this.

Thanks!

**Comment by ollo69** [comment]
> @ollo69, any chances we'll get this merged?

The real problem is that error 9006 seems to come continuously on API call, this is really different from error 9012 that was just coming every 30-45 minutes. I think that continue using this API is not anymore reliable. I'm not sure if make sense work on this integration to use new API because already exist native integration that use it!

**Comment by MPW1412** [comment]
Hi, thanks for your update. I hadn't know there are issues with the API. The usecase I try to solve is this: I want to set the machine to wash and dry and also set the temperature to 60 °C. So far, that's not possible with your add-on and according to my AI research, the native integration also doesn't support this. So this is not possible? The official LG App can do so, if I enable remote start. My goal is an automatic start when the solar panels produce enough energy.
